src/dbnode/client/write_state.go

```diff
@@ -49,17 +61,19 @@ type writeState struct {
 	sync.Mutex
 	refCounter
 
-	consistencyLevel                     topology.ConsistencyLevel
-	shardsLeavingCountTowardsConsistency bool
-	topoMap                              topology.Map
-	op                                   writeOp
-	nsID                                 ident.ID
-	tsID                                 ident.ID
-	tagEncoder                           serialize.TagEncoder
-	annotation                           checked.Bytes
-	majority, pending                    int32
-	success                              int32
-	errors                               []error
+	consistencyLevel                                    topology.ConsistencyLevel
+	shardsLeavingCountTowardsConsistency                bool
+	ShardsLeavingAndInitializingCountTowardsConsistency bool
+	topoMap                                             topology.Map
+	hostSucessMap                                       map[string]bool
```

Comment:
for small collections - maps are actually more expensive than slices because they mess up cache lines. do the following:

(1) benchmark slice v stack allocated array (say of size 4) v map for this collection
(2) don't de-reference ident.ID directly into `string`. use `ident.ID.Equals` for the comparisons

Reply:
also, 
- [ ] you need to update the `close()` method on the type to reset state and not leak memory. 
- [ ] add a test which validates the `close()` behaviour